File: engine-src/host/com/qwara/go/engine/PachiNative.java
// SPDX-FileCopyrightText: 2026 Qwara-chan
// SPDX-License-Identifier: GPL-3.0-or-later

package com.qwara.go.engine;

/**
 * Desktop-JVM copy of the Android bridge object, used to smoke-test libpachi.so on the host.
 *
 * <p>The app module supplies the real class with the same package and class name; only the four
 * native methods below matter, and they must stay in sync with the JNI bridge in
 * {@code engine-src/pachi_jni.cpp}.
 */
public final class PachiNative {
    static {
        System.loadLibrary("pachi");
    }

    private PachiNative() {}

    /** Start (or restart) the GTP protocol loop, resolving data files from {@code workDir}. */
    public static native void start(String workDir);

    /** Queue one GTP command line for the engine. */
    public static native void write(String line);

    /** Block until one engine output line is available; null once the session ended and drained. */
    public static native String readLine();

    /** Whether the GTP loop thread is still alive. */
    public static native boolean isRunning();

    /** Destroy the current session and join the engine thread. */
    public static native void shutdown();

    /**
     * Duplicate of the process stdout fd, taken before the first {@link #start} hijacks fd 1
     * for the engine. Host-side tests repoint System.out at it via /proc/self/fd/&lt;n&gt;.
     */
    public static native int dupStdout();
}

File: app/src/main/java/com/qwara/go/engine/PachiNative.kt
// SPDX-FileCopyrightText: 2026 Qwara-chan
// SPDX-License-Identifier: GPL-3.0-or-later

package com.qwara.go.engine

/**
 * Pachi 引擎 JNI 桥。libpachi.so 在 App 进程内以独立线程运行 GTP 协议循环，
 * 输入/输出经 dup2 管道重定向对接 Java 侧：
 *  - [start] 启动协议线程（幂等；EOF 退出后可再次调用以重建会话）；
 *  - [write] 写入一条 GTP 命令；
 *  - [readLine] 阻塞读取一行引擎输出，会话结束且队列耗尽后返回 null；
 *  - [isRunning] 协议线程是否存活；
 *  - [shutdown] 销毁当前会话并回收引擎线程；
 *  - [dupStdout] 会话劫持 fd 1 前复制进程 stdout（仅 host 校验器使用）。
 *
 * 协议行为与已知的坑见 engine-src/README.md（绝不发送 quit；EOF 结束会话；
 * 销毁顺序必须先恢复 stdio 再 join demux）。
 */
object PachiNative {
    init {
        System.loadLibrary("pachi")
    }

    external fun start(workDir: String)
    external fun write(line: String)
    external fun readLine(): String?
    external fun isRunning(): Boolean
    external fun shutdown()
    external fun dupStdout(): Int
}
